Added new kb article grid-dynamically-adjusting-column-header-styles (#2523)
telerik/blazor-docs: knowledge-base/grid-dynamically-adjusting-column-header-styles.md

---

> title: Dynamically Adjusting Column Header Cell Style in Blazor Grid

Reviewer: This title starts with two long words before we get to the important bit. Consider something shorter like 
```suggestion
title: Adjust Grid Column Header Styles Dynamically
```

description: Learn how to dynamically change the style of a column header cell in a Telerik Blazor Grid based on condition.
type: how-to

> page_title: How to Dynamically Style Column Header Cell in a Telerik Blazor Grid

Reviewer: ```suggestion
page_title: How to Style Grid Column Header Cells Dynamically in a Telerik Blazor Grid
```

slug: grid-dynamically-adjusting-column-header-styles

> tags: grid, blazor, header, headerclass

Reviewer: ```suggestion
tags: grid, blazor, headerclass
```

res_type: kb
ticketid: 1670074
---

## Environment

<table>
    <tbody>
	    <tr>
	    	<td>Product</td>

> 	    	<td>Grid for Blazor</td>

Reviewer: Perhaps the TreeList as well.

	    </tr>
    </tbody>
</table>

## Description

I am dynamically creating Grid columns in a loop and trying to adjust the column header cell style based on some condition and value from the [column field]({%slug components/grid/columns/bound%}#data-binding). Ideally, I want to implement logic to make this adjustment based on the value in the header cell. I have not been able to achieve this in the `HeaderTemplate` tag or in any of the cell render events handlers. In other templates, I have access to the `@context` field, which would make this easy, but that doesn't seem to work in the `HeaderTemplate`.

## Solution

> To dynamically add style to a column's header cell when dynamically creating columns in a loop, use the [`HeaderClass` parameter]({%slug components/grid/columns/bound%}#appearance) to set a class under a condition and apply different styles depending on the class. For scenarios with numerous and more complex conditions, you can create a method to determine the appropriate class.

Reviewer: Dynamically occurs twice

```suggestion
To style a column header cell when dynamically creating columns in a loop, use the [`HeaderClass` parameter]({%slug components/grid/columns/bound%}#appearance) to set a CSS class under a condition and apply different styles depending on the class. For scenarios with numerous and more complex conditions, you can create a method to determine the appropriate class.
```

Note that the [`HeaderTemplate`]({%slug grid-templates-column-header%}) does not receive a context argument because it is not related to rows and models, as outlined in the [Templates Overview]({%slug components/grid/features/templates%}) of the Telerik UI for Blazor documentation.

### Example

````CSHTML
<TelerikGrid Data="@GridData">
    <GridColumns>
        <GridColumn Field="@nameof(Product.Name)" Title="Product Name"/>
        @for (int i = 0; i <= MaxYears; i++)
        {

>             <GridColumn Field="@("Y"+i)" Title=@((StartYear + i).ToString()) DisplayFormat="{0:N}" HeaderClass="@GetHeaderClass(StartYear + i)"/>

Reviewer: ```suggestion
            <GridColumn @key="@i" Field="@( $"StartYear{i}" )" Title=@( $"{StartYear + i}" ) DisplayFormat="{0:N}" HeaderClass="@GetHeaderClass(StartYear + i)"/>
```

        }
    </GridColumns>
</TelerikGrid>

> <style>

Reviewer: We have one class for the future and two for the past. Consider simplifying or rewording "very-past".

>     .very-past-year-column {

Reviewer: ```suggestion
    .past-year-column {
```

        background-color: yellow;
    }

>     .past-year-column {

Reviewer: ```suggestion
    .previous-year-column {
```

        background-color: royalblue;
    }

    .current-year-column {
        background-color: pink;
    }

    .future-year-column {
        background-color: red;
    }
</style>

@code {

>     private List<Product> GridData { get; set; }

Reviewer: Avoid compiler null warnings:

```suggestion
    private List<Product> GridData { get; set; } = new();
```

or

```suggestion
    private List<Product>? GridData { get; set; }
```

    private int MaxYears = 10;

>     private int StartYear = 2020;

Reviewer: Use `{ get; set: }` or define these as constants.

    private string GetHeaderClass(int year)
    {
        switch (year)
        {
            case <= 2020:
                return "very-past-year-column";
            case < 2024:
                return "past-year-column";
            case 2024:

>                 return "current-year-column";

Reviewer: This checks will break in two months.

            default:
                return "future-year-column";
        }
    }

    protected override void OnInitialized()
    {
        GridData = new List<Product>();

>         Random rnd = new Random();

Reviewer: Use the static `Random.Shared.Next()` instead of `new Random().Next()`.
